util/expensivequery/expensivequery.go

```diff
@@ -73,12 +83,82 @@ func (eqh *Handle) Run() {
 				}
 			}
 			threshold = atomic.LoadUint64(&variable.ExpensiveQueryTimeThreshold)
+
+			instanceStats := &runtime.MemStats{}
+			runtime.ReadMemStats(instanceStats)
+			instanceMem := instanceStats.HeapAlloc
+			if err1 == nil && instanceMem > systemMem/10*8 {
+				// At least ten seconds between two recordings that memory usage is less than threshold (80% system memory).
+				// If the memory is still exceeded, only records once.
+				if time.Since(lastOOMtime) > 10*time.Second {
+					eqh.oomRecord(instanceMem, systemMem)
+				}
+				lastOOMtime = time.Now()
+			}
 		case <-eqh.exitCh:
 			return
 		}
 	}
 }
 
+func (eqh *Handle) oomRecord(memUsage uint64, systemMem uint64) {
+	logutil.BgLogger().Warn("The TiDB instance now takes a lot of memory, has the risk of OOM",
+		zap.Any("memUsage", memUsage),
+		zap.Any("systemMemoryTotal", systemMem),
+	)
+	sm := eqh.sm.Load().(util.SessionManager)
+	processInfo := sm.ShowProcessList()
+	pinfo := make([]*util.ProcessInfo, 0, len(processInfo))
+	for _, info := range processInfo {
+		if len(info.Info) != 0 {
+			pinfo = append(pinfo, info)
+		}
+	}
+	now := time.Now()
+
+	printTop10 := func(f func(i, j int) bool) {
+		sort.Slice(pinfo, f)
+		list := pinfo
+		if len(list) > 10 {
+			list = list[:10]
+		}
+		for i, info := range list {
+			logutil.BgLogger().Warn(fmt.Sprintf("OOM analyze SQL %v", i),
+				genLogFields(now.Sub(info.Time), info)...)
+		}
+	}
+
+	logutil.BgLogger().Warn("Top 10 memory usage of SQL for OOM analyze")
```

Comment:
How about print this table into a file along with the directory of `heap.profile`